rust/src/dataset.rs

```diff
@@ -1063,10 +1072,22 @@ pub(crate) async fn write_manifest_file(
         )
         .await?;
 
+    let id = uuid::Uuid::new_v4().to_string();
+    let staging_path =
+        Path::parse(format!("{path}-{id}")).map_err(|err| CommitError::OtherError(err.into()))?;
     // Then, create a copy at the latest manifest path
+    // While copy is usually atomic in object stores, copy is NOT atomic on *nix
+    // here we copy and rename to make sure both local and remote object stores
+    // have the same behavior
```

Comment:
It's still unclear to me what this is preventing. Could you explain what situation this prevents?

Reply:
This is try to prevent partially written `_latest.manifest` when running lance locally. Since we rely on `_latest.manifest` for resolving the latest version, a partially written `_latest.manifest` can cause errors.

Reply:
Got it. In a follow up I think we should create a wrapper method in `ObjectStore` that does this code path only on local filesystem, just so we aren't making too many IO calls.

Reply:
I'll make a issue in arrow-rs